modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/GmosControllerSim.scala

```diff
@@ -3,28 +3,49 @@
 
 package edu.gemini.seqexec.server
 
-import org.log4s._
+import java.util.concurrent.atomic.AtomicBoolean
 
+import org.log4s._
 import edu.gemini.seqexec.model.dhs.ImageFileId
 import GmosController.{GmosConfig, NorthTypes, SiteDependentTypes, SouthTypes}
 
-import scalaz.EitherT
+import scalaz.{EitherT, \/}
+import scalaz.Scalaz._
 import scalaz.concurrent.Task
 
 private class GmosControllerSim[T<:SiteDependentTypes](name: String) extends GmosController[T] {
   private val Log = getLogger
 
   override def getConfig: SeqAction[GmosConfig[T]] = ??? // scalastyle:ignore
 
+  private val stopFlag = new AtomicBoolean(false)
+  private val abortFlag = new AtomicBoolean(false)
+
+  private val tic = 200
+  private def observeTic(obsid: ImageFileId, remain: Int): Task[\/[SeqexecFailure, ImageFileId]] =
+    if(remain < tic) Task {
+      Log.info(s"Simulate Gmos $name observation completed")
+      TrySeq(obsid)
+    }
+    else Task.delay(stopFlag.get).flatMap( stop => Task.delay(abortFlag.get).flatMap( abort =>
+      if(stop) Task(TrySeq.fail(SeqexecFailure.Execution("Exposure stopped by user.")))
+      else if(abort) Task(TrySeq.fail(SeqexecFailure.Execution("Exposure aborted by user.")))
+      else Task{ Thread.sleep(tic.toLong)} *> observeTic(obsid, remain-tic)))
```

Comment:
I could have used a for comprehension, but I was worried about tail recursion.

Reply:
This isn't tail recursive anyway.  Add a `@tailrec` annotation (which is a good idea anyway).  The last branch of your `if/else` expression returns the result of `*>` which is passed the result of the recursive call.

Reply:
I'm not sure what you mean by being worried about tail recursion in this case